TSA: Scanner Gates and N-Spector update

DrDiasyl opened the pull request:
## About The Pull Request
Scanner gates now have a much-refined sprite with dirs.

https://github.com/user-attachments/assets/22659e12-5565-4feb-a8e7-0f1a56c4a99e

_The message in the video is outdated, now working with the formula `src.say("[detected_thing][reverse ? " not ": " "]detected!!")`_

Scanner gates now can have false positive/negative which depends on the tier of scanning modules in it:
- Tier 1 - 4%
- Tier 2 - 3%
- Tier 3 - 2%
- Tier 4 - 1%

Scanner gates now loudly blare into chat when they detect stuff.

Now there is a preset scanner gate to detect **GUNS** in main Brig entrances on maps, they are not upgraded with N-Spector.

***

N-Spector now can not only scan items for contraband but people too. It takes 4 seconds, makes a loud noise, and warns the person getting scanned in a chat with bold text.

https://github.com/user-attachments/assets/6aaca7bb-8273-485b-a727-c84f132b92f5

***

Made scanner code using proper cooldowns.

The scanner gate description now tells on which mode it's turned on.

Now smuggler satchel description tells you that it can prevent items from being detected by contraband scanners.
## Why It's Good For The Game
Scanner gates sprites were ancient and as such been updated.

Scanner gates having false-positive will make sure for Security to always be on their toes, and also cause interaction between crew who got false-positive. Also, it finally makes use of scanner gates being upgradable, as before this there was no reason for using higher-tier scanning modules for its construction.

Scanner gates blaring into the chat now make it clearer for spectators to acknowledge why the scanner got triggered.

Putting not upgraded scanners in the Brig entrance doors gives them some use outside of being sometimes built by bored Security Officers.

***

When N-Spector was upgraded, I immediately thought it should also be possible to scan people, like how TSA scans people with metal detectors.
It takes around the same time as stripping a backpack, and backpack searches still would be more effective as they can show stuff that the scanner cannot detect. 

The warning and loud sound make it possible for the person getting scanned to walk away if they start getting scanned unprompted. 
This would not likely cause any trouble, as going up and starting scanning someone for no reason would be the same as going up to someone and starting to strip their backpack.

With N-Spector Security can perform more fluff interesting searches. 
## Changelog
:cl: DrDiasyl aka DrTuxedo
add: Scanner gates now blare into the chat the reason why it got triggered
add: Scanner gates now can have false positives/negatives, and the chance of them being reduced when they are upgraded
add: Brig entrances now have scanner gates preset to detect GUNS
add: You can now scan people with N-Spector for contraband
image: Scanner gates now have a better sprite with dirs
qol: Scanner gates description now tells to what mode they are set
fix: You no longer can remove N-Spector from scanner gates without unlocking them first
/:cl:

DrDiasyl committed:
TSA update

DrDiasyl committed:
Guns Scanners to Brig enterances

DrDiasyl committed:
False Positive

Watermelon914 commented:
Those false positive percentages might be too high. If placed in a place with high travel density, it'd trigger every 1/20 times on T3 parts. You might want to decrease the percentages to be much lower so that it's still possible to occur but generally more rare.

Watermelon914 commented on code/game/machinery/scanner_gate.dm:
```diff
@@ -46,6 +46,10 @@
 	///Does the scanner ignore light_pass and light_fail for sending signals?
 	var/ignore_signals = FALSE
+	///Modifier to the chance of scanner being false positive
+	var/false_positive = 0
```
this variable name is misleading. If this intends to be a reduction to the chance of a false positive, it needs to be made more clear.
The better thing to do might be to negate the value instead and then add it to the base_false_positive value when doing a prob check

Watermelon914 commented on code/game/machinery/scanner_gate.dm:
```diff
@@ -73,6 +82,8 @@
 /obj/machinery/scanner_gate/examine(mob/user)
 	. = ..()
+
+	. += span_notice("It's set to scan for [span_boldnotice(scangate_mode)]")
```
This information doesn't need to be revealed unless unlocked, imo

DrDiasyl replied:
It is already visible when unlocked, this is why I made it also visible when not unlocked.

I want antags and such to know what this scanner detects, and if they should fear it (Weapons, Contraband) or not (Infections, Nutrition)

OverwatchVoice commented:
I love this! One thing I'd suggest is an emag interaction that wrecks the scanner gate somehow, like never displaying positives

Sadboysuss commented on code/game/machinery/scanner_gate.dm:
```diff
@@ -251,12 +276,15 @@
+		return
+
+	if(detected_thing)
+		src.say("[reverse ? "Not" : "A"] [detected_thing] person detected!!")
```
instead of saying what "kind" of person they are: say what the scanner detected, Example: "Romerol infection detected" or "Weapons detected"

DrDiasyl replied:
The scanner does only scan carbons so why though? Its always a person

Sadboysuss replied:
if it's always a person then why clarify it?

DrDiasyl committed:
clearer var + lower chance

Base false positive chance is now from 8 to 5.

DrDiasyl commented:
> Those false positive percentages might be too high. If placed in a place with high travel density, it'd trigger every 1/20 times on T3 parts. You might want to decrease the percentages to be much lower so that it's still possible to occur but generally more rare.

Updated the chances.
**OLD:**
- Tier 1 - 7%
- Tier 2 - 6%
- Tier 3 - 5%
- Tier 4 - 4%

**NEW:**
- Tier 1 - 4%
- Tier 2 - 3%
- Tier 3 - 2%
- Tier 4 - 1%

Sadboysuss commented:
You should make the smuggler bag bypass this scanner, so it actually has a use

DrDiasyl commented:
> You should make the smuggler bag bypass this scanner, so it actually has a use

It already does.

atlasle commented:
smallest nitpick but could "a armed person detected" just be "armed person detected"

DrDiasyl committed:
Clarity message

grungussuss reviewed (commented):
some minor spellcheck nits

DrDiasyl commented on code/game/objects/items/inspector.dm:
```diff
@@ -8,7 +8,7 @@
 /obj/item/inspector
 	name = "\improper N-spect scanner"
-	desc = "Central Command standard issue inspection device. Can perform either wide area scans that central command can use to verify the security of the station, or detailed scans to determine if an item is contraband."
+	desc = "Central Command standard issue inspection device. Can perform either wide area scans that central command can use to verify the security of the station, or detailed scan. Can scan people and items for being contraband."
```
I would say that antags are a contraband L

DrDiasyl committed:
spelling

DrDiasyl committed:
Fixed oversight of N-Spector in Gates

Now you cannot just pop out the N-Spector out of Scanner Gates willy nilly, you have to unlock them first (done by emag or proper ID)

CRITAWAKETS commented:
what stops someone with a false positive from just... walking through the gate again?

DrDiasyl commented:
> what stops someone with a false positive from just... walking through the gate again?

Good point, might wanna add false negatives as well

DrDiasyl committed:
false negatives!

Wallemations reviewed (changes requested):
The side sprites look a bit odd with their perspective, I think you need more top and less sides

DrDiasyl commented:
> The side sprites look a bit odd with their perspective, I think you need more top and less sides

@Wallemations like this?
![image](https://github.com/user-attachments/assets/511b2fb8-6aef-4e8d-a5c0-0be238702e9a)

DrDiasyl committed:
better prespective

Wallemations commented:
> > The side sprites look a bit odd with their perspective, I think you need more top and less sides
> 
> 
> 
> @Wallemations like this?
> 
> ![image](https://github.com/user-attachments/assets/511b2fb8-6aef-4e8d-a5c0-0be238702e9a)
> 
> 

Certainly better

DrDiasyl commented:
well, already committed it

Wallemations reviewed (approved):
Sprites look good

Wallemations commented:
Due to the amount of code changes I'll leave it to a code maint to merge

Mothblocks commented on code/game/machinery/scanner_gate.dm:
```diff
@@ -234,8 +267,15 @@
 		beep = !beep
+
+	if(prob(base_false_beep - minus_false_beep)) //False positive/negative
+		if(prob(50))
+			beep = TRUE
+		else
+			beep = FALSE
```
```suggestion
		beep = prob(50)
```

Mothblocks commented on code/game/objects/items/inspector.dm:
```diff
@@ -106,18 +107,27 @@
+		if(!COOLDOWN_FINISHED(src, scanning_person))
+			return ITEM_INTERACT_BLOCKING
+
+		visible_message(span_warning("[user] starts scanning [interacting_with] with [src]"))
+		to_chat(interacting_with, span_userdanger("[user] is trying to scan you for contraband!"))
```
balloon alerts instead

DrDiasyl replied:
why balloon alerts? Userdanger is big red text and it is used for example when you're getting stripped or attacked aka it is noticeable

Mothblocks commented on code/game/machinery/scanner_gate.dm:
```diff
@@ -251,12 +291,15 @@
+		return
+
+	if(detected_thing)
+		src.say("[detected_thing][reverse ? " not " : " "]detected!!")
```
```suggestion
		say("[detected_thing][reverse ? " not " : " "]detected!!")
```

Mothblocks commented on code/game/machinery/scanner_gate.dm:
```diff
@@ -73,6 +82,8 @@
 /obj/machinery/scanner_gate/examine(mob/user)
 	. = ..()
+
+	. += span_notice("It's set to scan for [span_boldnotice(scangate_mode)]")
```
```suggestion
	. += span_notice("It's set to scan for [span_boldnotice(scangate_mode)].")
```

Mothblocks commented on code/game/machinery/scanner_gate.dm:
```diff
@@ -321,6 +364,11 @@
 						detect_nutrition = NUTRITION_LEVEL_FAT
 			. = TRUE
 
+/obj/machinery/scanner_gate/preset
```
I don't really understand what "preset" means in this context. If its just guns, would be better named `scanner_gate/preset_guns` or something

DrDiasyl committed:
what mothblocks asked

DrDiasyl committed:
Delete scanner.txt

Mothblocks commented on code/game/objects/items/inspector.dm:
```diff
@@ -106,18 +107,28 @@
+
+		visible_message(span_warning("[user] starts scanning [interacting_with] with [src]"))
+		to_chat(interacting_with, span_userdanger("[user] is trying to scan you for contraband!"))
+		balloon_alert_to_viewers("scanning")
```
```suggestion
		balloon_alert_to_viewers("scanning...")
```

Mothblocks committed:
Ellipsis
